package typtop;

message Log {
    required bool  in_cache = 1;
    required bool  istop5fixable = 2;
    required int32 edit_dist = 3;
    required float rel_entropy = 4;
    required float pass_complexity = 5;
    required uint32 tid = 6;
    required int64 ts = 7;
    required string localtime = 8;
}

message ConfigHeader {
    optional bool  allowed_typo_login = 1 [default = true];
    optional int32 period_for_sending_log = 2 [default = 21600]; // 6 hours
    optional int32 typo_expiry_time = 7 [default = 864000];  // expire a typo if not used in 10 days
    optional bool  allow_upload = 3 [default = true];
    required string install_id = 4;
    required bytes public_key = 5;
    required bytes global_salt = 6;
}

enum SystemStatus {
    ALL_GOOD = 0;
    PW_CHANGED = 1;
    UNINITIALIZED = 3;
};

message EncHeaderData{
    required bytes pw = 1;
    repeated int32 freq = 2;
    repeated int64 last_used = 3;   // expire a typo if it is too old
    required float pw_ent = 4;
}
message Header {
    optional int32 indexj = 1 [default = 0];
    optional int32 last_log_sent_time = 2 [default = 0];
    optional int32 login_count = 3 [default = 0];
    optional SystemStatus sys_state = 4 [default = UNINITIALIZED];
    required bytes enc_header = 5;
}

message Logs { // for sending the log to the server
    repeated Log L = 1;  // Log
}

message typoDB {
    optional string version = 9 [default = "1"]; // the version of the db file.
    required ConfigHeader CH = 1; // Config Header
    required Header H = 2;  // Header
    repeated bytes W = 3; // Waitlist
    repeated bytes T = 4; // TypoCache
    optional Logs logs = 5;  // Log
}


// Waitlist contains encrypted version of pw and time
message WaitlistEntry {
    required bytes pw = 1;
    required int64 ts = 2;
}
